pkg/runner/chroot_runner.go

```diff
@@ -0,0 +1,73 @@
+package runner
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type chrootRunner struct {
+	base               Runner
+	buildDirectoryPath string
+}
+
+// NewChrootRunner modifies the arguments of the incoming RunRequest so
+// that the first argument is an absolute path resolved using the path
+// in the request using the input root as the chroot
+func NewChrootRunner(base Runner, buildDirectoryPath string) Runner {
+	return &chrootRunner{
+		base:               base,
+		buildDirectoryPath: buildDirectoryPath,
+	}
+}
+
+func findExecutable(file string) error {
+	d, err := os.Lstat(file)
+	if err != nil {
+		return err
+	}
+	if m := d.Mode(); !m.IsDir() && m&0111 != 0 {
+		return nil
+	}
+	return os.ErrPermission
+}
+
+func (r *chrootRunner) lookPathChroot(request *runner.RunRequest) (string, error) {
+	file := request.Arguments[0]
+	if strings.Contains(file, "/") {
+		if err := findExecutable(file); err != nil {
+			return file, err
+		}
+		return file, nil
+	}
+	for _, dir := range filepath.SplitList(request.EnvironmentVariables["PATH"]) {
+		if dir == "" {
```

Comment:
I think `filepath.Join()` already does the right thing when the string is empty.